### ISSUE TITLE
[Strix 09/20] Add a fak-native HIP long-prompt prefill leaf for Qwen3.8 on Strix Halo

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12539]
# Add a fak-native HIP long-prompt prefill leaf for Qwen3.8 on Strix Halo

Parent campaign: #11572. Frozen comparison ladder: #12093. Matched external-comparison gate: #12098.

## Context

- For: Strix Halo users whose Qwen3.8 workload is dominated by 2K–32K prompt ingestion.
- Problem: fak has no physically accepted native HIP sequence-prefill panel for the Qwen3.8 full-attention geometry.
- Today: the recovered three-file candidate is uncommitted and independently refuted as-is: it admits head dimension 128 instead of the authoritative 24 Q heads / 4 KV heads / head dimension 256, its lexical test misses that contradiction, and the published physical command names selectors that trunk does not support.
- Better because: one narrow, default-off HIP component can test AMD-native resident launch/runtime behavior without silently changing the serving backend or claiming a full ROCm/model path.
- Witness: the corrected source contract passes locally, then the exact candidate bytes compile and run through #12597 and receive the repeated physical component verdict owned by #12598.

- Centrality: Core
- P1 Context: advanced — targets the largest observed AMD-native prefill gap at the exact Qwen3.8 attention geometry while keeping unrelated serving state out of scope.
- P2 Net value: advanced — requires resident-tensor counters plus a matched physical Vulkan/HIP receipt before the candidate may land or claim value.
- P3 Adaptation: preserved — the backend remains build-tagged, explicitly opt-in, default-off, and incapable of substituting CPU, Vulkan, or llama.cpp.
- P4 Operations: advanced — source, compiler, device, workload, lease, timeout, transfer, fallback, and keep/reject state are explicit and fail closed.

## Current state

HOLD before commit. The original candidate is preserved read-only at base `9121ca5e9cc78a8292c9898e5e0137a57758fe3a` with combined name/content SHA-256 `6bc960c884fdc2bfeaab83f7fe19cefe9599975f9234f41a95543bb921ceb19b`. Recovery may correct the three source files as an executable disjoint slice, but no commit, closure, or performance claim is allowed until the exact corrected bytes receive the required physical receipt.

The sanctioned target remains the physical gfx1151 Strix appliance. Availability of the required HIP-language/offload compiler and runtime path is not yet witnessed. #12597 owns the public, source-bound validator route; absence of that route is a typed HOLD, never a software pass.

## Parent context

Parent: #11572

## Dependencies and pickup rule

- Coordinates with: #266
- Coordinates with: #12597
- Coordinates with: #12598
- Coordinates with: #12540
- Promotion requires: #12598
- Promotion requires: #12098

There is no Start-blocked relation because correction of the existing three-file source candidate is independently executable. Landing and closure are held by #12598's exact-byte physical receipt; #12098 additionally gates external comparison claims. Full-model HIP-to-Vulkan state handoff and continuation belong to #12540, not this component leaf.

## Scope

Recover and correct exactly one default-off gfx1151 HIP causal-attention prefill panel in these three files:

- `internal/compute/rocm.go`
- `internal/compute/rocm_qwen38_prefill.hip`
- `internal/compute/rocm_source_contract_test.go`

The Go adapter must carry the `linux && rocm && cgo` build tag, deterministic gfx1151 HIP build directive, explicit opt-in registration, flat resident-tensor ABI, checked allocation/geometry, and transfer/kernel/fallback counters. The HIP unit must implement the exact Qwen3.8 full-attention geometry (24 Q heads, 4 KV heads, head dimension 256), keep transfers outside the measured panel, and fail closed outside the admitted geometry. The source witness must inspect function-local contracts and reject geometry, ABI, counter, registration, launch, or fallback mutations.

## Core through-line

Recover exact three-file source -> correct Qwen3.8 geometry and safety -> prove the non-vacuous source contract -> compile/execute the exact bytes through #12597 -> consume #12598's physical keep/reject receipt -> land only a KEEP candidate.

## Gold-plating boundary

No general ROCm backend, decode path, model loader, quantized weights, full-model logits/KV/greedy continuation, HIP-to-Vulkan handoff, multi-GPU work, CUDA port, competitor claim, or hidden fallback.

## Done condition

The issue is done in exactly one of two witnessed states:

1. KEEP: the corrected three files pass the deterministic software witness, compile for gfx1151, and #12598 records physical component parity plus the repeated pp2048/pp32768 keep gate for those exact bytes; one guarded commit then lands only the three files.
2. REJECT: the physical receipt records a correctness or performance rejection, Vulkan remains selected, the candidate is not committed, and the preserved WIP is dispositioned without a HIP performance claim.

Source-text success alone is not done and is not permission to commit.

## Witness

Software contract:

```text
go test -count=1 -run '^TestROCmQwen38PrefillSourceContract$' ./internal/compute
```

Physical component gate after #12597:

```text
fak-dev amd-strix-validate --subkernels=qwen38-long-prefill --ablate=none --evidence-only --json
```

#12598 owns one warm-up plus at least five measured Vulkan/HIP repetitions at pp2048 and pp32768, component-output parity, finite output, per-run/median timing, `engine=fak-native`, zero measured transfers, and `fallback_count=0`. Its test owns both arms; the former invalid `--ablate=vulkan,hip` command is retired. Candidate changes invalidate the physical review and receipt.

## Likely files

- `internal/compute/rocm.go`
- `internal/compute/rocm_qwen38_prefill.hip`
- `internal/compute/rocm_source_contract_test.go`

## Expected steps

5

## Work estimate

Estimate: 3 points.

## Overall completion contribution

Contribution: 3/100 points.

## Completion standard

experiment

## Target operating envelope

- physical gfx1151 executions: >= 24 runs
- prompt lengths covered: >= 2 lengths
- measured repetitions per arm and length: >= 5 runs
- component parity pass rate: = 100 percent
- finite output pass rate: = 100 percent
- measured transfer count: = 0 transfers
- backend fallback count: = 0 calls
- best-length HIP to Vulkan median speed ratio: >= 1.05 ratio
- other-length HIP to Vulkan median speed ratio: >= 0.98 ratio

## Witnessed operating envelope

- physical gfx1151 executions: = 0 runs
- prompt lengths covered: = 0 lengths
- measured repetitions per arm and length: = 0 runs
- component parity pass rate: = 0 percent
- finite output pass rate: = 0 percent
- measured transfer count: = 0 transfers
- backend fallback count: = 0 calls
- best-length HIP to Vulkan median speed ratio: = 0 ratio
- other-length HIP to Vulkan median speed ratio: = 0 ratio

### Work unit

leaf

## Closure binding

Any KEEP commit must cite #12539, carry `(fak compute)`, contain exactly the three likely files, and be bound to both a green software test and #12598's exact-byte physical receipt. A REJECT closes by witnessed issue disposition with no code commit.

**Comment by anthony-chaudhary** [comment]
Claimed for the bounded three-file gfx1151 HIP Qwen3.8 prefill leaf. Pickup gate checked: no Start blocked by relation; #12098 remains promotion-only. I will preserve engine=fak-native and fallback_count=0 and report software plus Strix witness receipts.

**Comment by anthony-chaudhary** [comment]
Implementation candidate is complete within the declared three-file boundary, but I am not landing it because the required physical witness is structurally unavailable at the current trunk.

Confirmed candidate paths only:
- `internal/compute/rocm.go`
- `internal/compute/rocm_qwen38_prefill.hip`
- `internal/compute/rocm_source_contract_test.go`

Software evidence:
- `go test -count=1 -run TestROCmQwen38PrefillSourceContract ./internal/compute` => PASS.
- Isolated overlay build/vet/test completed successfully during the Strix validation attempt.
- The candidate remains default-off, gfx1151-only, `engine=fak-native`, resident-only in the measured ABI, and enforces `fallback_count=0`.

Required hardware command evidence:
- The exact command from the issue was invoked; on this peer-dirty checkout it first requires explicit `--mine` ownership.
- Re-run with the three declared `--mine` paths completed overlay build/vet/test, then failed closed before device dispatch: `amdgpu: unknown subkernel selector(s): qwen38-long-prefill`.
- Current `internal/amdgpu/strix_subkernels.go` has no spec or parity-contract row for that selector. Adding it would exceed this issue's authoritative three-file write boundary.
- The sanctioned `strix-agent` route is reachable, but its current execution profile has no `hipcc`, so a direct gfx1151 HIP compile could not substitute for the missing registered validator path.

Prior-art correction: `peonist-ai/halogen-server@6a51c35` publishes the gfx1151/no-fallback performance hypothesis and binaries/docs, not reusable HIP kernel source. The candidate is therefore independently implemented and marked `INSPIRE-ONLY`; no Halogen kernel source was copied.

Next checkable step: reconcile the issue contract so `qwen38-long-prefill` is admitted by the Strix selector/parity registry and the selected Strix execution profile provides `hipcc`; then rerun the exact pp2048/pp32768 N>=5 Vulkan/HIP witness. Until that passes, no commit/push and no promotion claim.

**Comment by anthony-chaudhary** [comment]
Resumed the bounded three-file recovery under an exact-path lease. The deleted original worktree candidate is recoverable from its saved creation and correction patches. Correcting the exact 24 Q / 4 KV / 256 head geometry and function-local source witness in an isolated checkout. #12598 owns physical parity and the #12597 prerequisite route; no commit, closure, deployment, or performance claim will precede the exact-byte physical KEEP/REJECT receipt.

**Comment by anthony-chaudhary** [comment]
Coordination from dedicated #12598 witness task: inspecting generic #12597 fresh-build route and independent device test. Please publish the corrected three-file candidate location/digest and final Go/flat-ABI names for exact-byte witness integration. No hardware window acquired here. Current trunk also has rocm_backend.go and libfakrocm core/Qwen units, which the fresh build must link alongside this candidate; preserve independent component registration/default-off behavior.

**Comment by anthony-chaudhary** [comment]
Source recovery now contains the exact 24 Q / 4 KV / 256 geometry, shared ROCm serialization and one-time buffer lifetime checks. Coordination with #12598 added a package-private preallocated-output adapter inside the same three-file boundary, with input/output ownership checks and output-alias rejection before the HIP ABI. Independent source review approved the final adapter and mutation-sensitive contract. An earlier software PASS report was refuted by direct execution (generic-head mutation escaped); the contract was repaired and that intermediate revision passed with actual stdout and exit 0. The final adapter revision is undergoing a fresh direct-exit WSL witness; no final software or hardware PASS is claimed yet. Candidate remains uncommitted, default-off, and held for #12597/#12598 exact-byte physical qualification. No performance or full-model serving claim.

**Comment by anthony-chaudhary** [comment]
Final corrected three-file software candidate is ready for #12597/#12598 qualification. The direct WSL witness `go test -count=1 -run ^TestROCmQwen38PrefillSourceContract$ ./internal/compute` completed with actual process exit 0 and package result ok (0.022s). Independent exact-byte source review passed. The adapter supplies the allocation-free preallocated-output seam requested by #12598, rejects output/input aliasing, and reports the architecture observed and checked by HIP. Final SHA-256: rocm.go 2ac14bdf2dbc7d135d09df6285b6de115bacca76f028f6aaacbf84bcb94c71a9; rocm_qwen38_prefill.hip f82b9101f5b22894b0cd6283ca18cbcd2e79d6d6ceb4749e76b18fad6f17f019; rocm_source_contract_test.go 1623305c82597dadcbbff58a9d2e3597988fcfb35653e893fd7230e48b48c8d1. Base f1bcc34e0a4962231d1ee5d72364c22accd2c1c7. These receipts prove software source contracts only. No HIP compilation, physical execution, parity, or speed acceptance is claimed. Candidate remains uncommitted and issue OPEN pending the exact-byte physical KEEP/REJECT gate owned by #12598; no full-model claim.

**Comment by anthony-chaudhary** [comment]
Integration follow-up requested by #12598 invalidates the prior freeze: adding observed HIP PCI domain:bus:device.function identity within the same three files. D2D observability was investigated: no runtime instrumentation currently exists in this component. Absence of current D2D calls will not be represented as measured zero. The existing four-counter launch ABI remains unchanged; a separate explicit unsupported-observability capability will let the consumer fail closed when D2D evidence is required. No hardware execution or commit authorized by this change; new hashes and software review will follow.

**Comment by anthony-chaudhary** [comment]
Accepted the parent-approved linker correction requested by #12598: this optional component will additionally require the rocm_qwen38_prefill build tag, preserving ordinary rocm core builds without the candidate archive. #12598 dual profile/device test should use vulkan,rocm,rocm_qwen38_prefill. Identity seam returns validated canonical observed PCI BDF; PrefillD2DBytesObserved() reports false because runtime D2D instrumentation is absent. The four-counter launch ABI is unchanged. This is an explicit unsupported D2D capability, never measured zero. Final focused software witness and new hashes pending; no hardware or landing claim.

**Comment by anthony-chaudhary** [comment]
New frozen source-only handoff after PCI identity/linker integration: independent source review PASS and coordinator direct WSL witness `go test -count=1 -run ^TestROCmQwen38PrefillSourceContract$ ./internal/compute` actual exit 0, stdout ok compute (0.041s). Final SHA256: rocm.go bc5af5ed92e35000df645d75e56d1195f5a0bbab8689c55333fbb076f60abba4; rocm_qwen38_prefill.hip 41974e5b82c1256362e1e68e47c90a945c37233093b35d34bbf5631a8ddcccdd; rocm_source_contract_test.go aeab6c3f5d6bc23e41cefedcd70da246e4981c37d2fa87c6c0ead84c6c25c8b2. Requires linux && rocm && rocm_qwen38_prefill && cgo, preserving generic ROCm build linkage. PrefillPhysicalDeviceIdentity() returns strictly validated canonical observed HIP PCI BDF. PrefillD2DBytesObserved() returns false: runtime D2D instrumentation is unsupported, so consumers must fail any required D2D-observation gate; no measured-zero claim. Four-counter launch ABI unchanged. All previous source digests superseded. Exactly three uncommitted candidate files remain; no HIP tagged compile, physical acceptance, or performance credit. #12598 consumes this source for its independently governed build/physical gate; issue remains OPEN.
